### PR TITLE
Start standby master web server

### LINKED ISSUE #13840: Non Leader master should also start web server
**Is your feature request related to a problem? Please describe.**

- When ha didn't work as expected, or the leader didn't transfer to a new master, there are no leader that time, so we want to the the stacks from web page.

- We want to scrap the metrics of non-leader masters by prometheus, to monitor the gc, heap, and some other metrics of it.

**Describe the solution you'd like**

Start the web server all the time, although it isn't leader.

**Describe alternatives you've considered**

Start the web server all the time, although it isn't leader.

**Urgency**
High

**Additional context**
No

### PR DESCRIPTION
**Pull request by bzheng888**
### What changes are proposed in this pull request?
Start web server for standby master, FIx #13840 
1. In FaultTolerantAlluxioMasterProcess.start(), move the startWebServing() function before waiting raft leader selection loop. 
2. In AlluxioMasterProcess.start(), add startWebServing() to keep none Ha case work.
3. Remove startWebServing() from  AlluxioMasterProcess.startServing().
4. Remove stop web server logic from AlluxioMasterProcess.stopServing.
5. Add stop web server logic to AlluxioMasterProcess.stop().
### Why are the changes needed?
It help us get some useful  metric info from standby master.
Please clarify why the changes are needed. For instance,
No
### Does this PR introduce any user facing changes?
No
Please list the user-facing changes introduced by your change, including
No

### CODE REVIEW

**Comment by maobaolong** [comment]
@bzheng888 Would you like to share your tests result?

**Comment by LuQQiu** [comment]
@bzheng888 can you help to make the standby master web server a configurable option and by default not enabling similar to fuse process? https://github.com/Alluxio/alluxio/pull/13892

**Review by jiacheliu3 — COMMENTED** [review]
When your current primary master loses primacy, what is the behavior on the web server? Vice versa when your secondary gains primacy what is the behavior?

**Comment by bzheng888** [comment]
> @bzheng888 can you help to make the standby master web server a configurable option and by default not enabling similar to fuse process? #13892

Ok, I will add the configuration

**Comment by bzheng888** [comment]
> @bzheng888 can you help to make the standby master web server a configurable option and by default not enabling similar to fuse process? #13892

@LuQQiu Thanks, I will refer to this pr and add the configuration.

**Comment by bzheng888** [comment]
> When your current primary master loses primacy, what is the behavior on the web server? Vice versa when your secondary gains primacy what is the behavior?
@jiacheliu3 Thanks for reviewing this, In this pr logic, the web server will be started before the node becoming primary or secondary, and no matter the raft node's state changes, the web server is running all time until the master process is stopped. Further more, I will make this configurable and the default configuration will stop this behavior, does this have a problem?

**Comment by bzheng888** [comment]
> When your current primary master loses primacy, what is the behavior on the web server? Vice versa when your secondary gains primacy what is the behavior?

@jiacheliu3 Thanks for reviewing this, In this pr logic, the web server will be started before the node becoming primary or secondary, and no matter the raft node's state changes, the web server is running all time until the master process is stopped. Further more, I will make this configurable and the default configuration will stop this behavior, does this have a problem?

**Comment by maobaolong** [comment]
@jiacheliu3 I think it can make sense to open the secondary master web service but not include write possible operation REST api, just like hdfs standby namenode.

About the metrics in secondary master, it can be right value if we don't think the current metrics stand for the cluster master metrics. The value of secondary master include two part, first part cannot be changed because only primary master can change these kind of metrics, e.g. TotalRpcCount, rpcCallqueueLength. second part stand for the state of this master, something like, jvm pause time, thread count, heap usage, current journal sync sequence index and so on.

Further more, leverage this step, we can display the following useful feature
- Display the leadership of each master
- Display all the election information, something like who vote to the leader, leadership switch timestamp, vote cost time.
- Display the current journal sequence index.

**Comment by maobaolong** [comment]
@bzheng888 As we have a offline discussion between @bzheng888 @jiacheliu3 @maobaolong , would you like to paste the consensus conclusion here?

**Comment by bzheng888** [comment]
> @bzheng888 As we have a offline discussion between @bzheng888 @jiacheliu3 @maobaolong , would you like to paste the consensus conclusion here?

I think the pr 13927 is more complete and it gives a good idea to redirect some request to primary master, I think it's better to finish design doc based on that. There are two problem i'm concerned:
1.  Should we still add a configuration to enable the standby master web server?
2. I think its necessary to show whether the master is leader or non-leader.
Finally, I will post this problem in the design doc, it will be done in one or two days.

**Comment by qian0817** [comment]
I think it should't stop jvm pause monitor and metrics system.

**Comment by bzheng888** [comment]
https://docs.google.com/document/d/1dKWLbrT926vCk5vSAmAIbcjvppGLddl1DfAjAzqMBhQ/edit?usp=sharing

**Comment by jiacheliu3** [comment]
> @maobaolong @jiacheliu3, here is the design doc draft, please take a a look and give me some advice
> https://docs.google.com/document/d/1dKWLbrT926vCk5vSAmAIbcjvppGLddl1DfAjAzqMBhQ/edit?usp=sharing

Hi @bzheng888 could you open the permission to everyone that has the link? Thanks!

**Comment by bzheng888** [comment]
> > @maobaolong @jiacheliu3, here is the design doc draft, please take a a look and give me some advice
> > https://docs.google.com/document/d/1dKWLbrT926vCk5vSAmAIbcjvppGLddl1DfAjAzqMBhQ/edit?usp=sharing
> 
> Hi @bzheng888 could you open the permission to everyone that has the link? Thanks!

Done, please have a try again.

**Comment by jiacheliu3** [comment]
@apc999 @LuQQiu  FYI

**Comment by LuQQiu** [comment]
Reviewed design doc

**Comment by ggezer** [comment]
Will we start serving REST calls from stand-by masters with this change ?!?

**Comment by bzheng888** [comment]
> Will we start serving REST calls from stand-by masters with this change ?!?

No ,  will not change REST api!

**Comment by maobaolong** [comment]
@apc999 @LuQQiu @jiacheliu3 @ggezer @bzheng888 
Thanks for the reviewing and suggestion for this PR.

As we have a discussion in today's Alluxio FUSE & AI community sync meeting, the following we have reached consensus.
- For this PR, we just start secondary master web server.
- There are some follow-up task after this PR
  - Discuss two approach to implement a global view web service.
    - Approach A: Redirect secondary web request to primary master
    - Approach B: Start a new service, let's call it `alluxio-monitor`, all the web request for this server will redirect to the primary master, meanwhile, the secondary master should do nothing just display itself state through web server.
  - Improve the design doc, add goal and non-goal for this PR
  - Create a new issue to track the `global view web service` with a new design doc.
- We can discuss the follow-up task to get more detail by a new meeting.

Please left comment if the conclusion is not correct.

**Comment by ggezer** [comment]
> > Will we start serving REST calls from stand-by masters with this change ?!?
> 
> No , will not change REST api!

I was not asking whether you change it or not. By starting a webserver, you're also exposing REST API from secondaries.

**Comment by ggezer** [comment]
> @apc999 @LuQQiu @jiacheliu3 @ggezer @bzheng888
> Thanks for the reviewing and suggestion for this PR.
> 
> As we have a discussion in today's Alluxio FUSE & AI community sync meeting, the following we have reached consensus.
> 
> * For this PR, we just start secondary master web server.
> * There are some follow-up task after this PR
>   
>   * Discuss two approach to implement a global view web service.
>     
>     * Approach A: Redirect secondary web request to primary master
>     * Approach B: Start a new service, let's call it `alluxio-monitor`, all the web request for this server will redirect to the primary master, meanwhile, the secondary master should do nothing just display itself state through web server.
>   * Improve the design doc, add goal and non-goal for this PR
>   * Create a new issue to track the `global view web service` with a new design doc.
> * We can discuss the follow-up task to get more detail by a new meeting.
> 
> Please left comment if the conclusion is not correct.

What's the purpose of "global view web service" ?

**Comment by maobaolong** [comment]
> What's the purpose of "global view web service" ?

@ggezer The purpose was first proposed by @jiacheliu3 , aim to supply a global alluxio web service, no matter which master you accessed, primary or secondary, the view and result are the same. In fact I can not totally support this idea, I think each master (primary or secondary) should display itself state, base on this viewpoint, I proposed approach B, if we want to supply a global alluxio cluster web service, we can invent a new service "alluxio monitor", all the web request will redirect to primary master.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 1]
```diff
@@ -150,13 +150,21 @@ public InetSocketAddress getRpcAddress() {
   public void start() throws Exception {
     mJournalSystem.start();
     mJournalSystem.gainPrimacy();
+    if (ServerConfiguration.getBoolean(PropertyKey.SECONDARY_MASTER_WEB_ENABLED)) {
```
Comment (maobaolong) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:153]: This seems can be reached only when the master is leader?

**Comment by bzheng888** [comment]
> > > Will we start serving REST calls from stand-by masters with this change ?!?
> > 
> > 
> > No , will not change REST api!
> 
> I was not asking whether you change it or not. By starting a webserver, you're also exposing REST API from secondaries.

@ggezer Sorry about misunderstanding your question, I think exposing the REST API has no effect,  can you tell me what your worries are?

**Comment by LuQQiu** [comment]
REST API just like web pages which the leading master serves real requests. Standby master should not serve any requests through REST API, so REST should be closed on secondary master

**Comment by LuQQiu** [comment]
@maobaolong @bzheng888 @jiacheliu3 @ggezer 
Based on my understanding, the approaches @maobaolong mentioned are big scope and need really careful discussion. 
We agreed to go for the simplest approach:
- Support launching secondary master web server which shows all the information belonging to this master (logs, metrics, etc)
- for all the information that doesn't belong to this server, like the leading master information and REST service, we don't show/expose.
- There should be an easy way to show whether this master is a leading master or not.

There should be multiple PRs to achieve this approach, we can open a feature branch for it.

**Comment by jiacheliu3** [comment]
@maobaolong close this if we are moving to https://github.com/Alluxio/alluxio/pull/14259 ?

**Comment by maobaolong** [comment]
@jiacheliu3 Yeah, thanks for remind us.

**Pull request closed without merging**